lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp

```diff
@@ -631,6 +637,21 @@ LifetimeChecker::LifetimeChecker(const DIMemoryObjectInfo &TheMemory,
   // locally inferred by the loop above.  Mark any unset elements as not
   // available.
   MemBBInfo.setUnknownToNotAvailable();
+
+  // Finally, check if we need to emit compatibility diagnostics for cross-module
+  // non-delegating struct initializers.
+  if (TheMemory.isStructInitSelf() &&
+      !Module.getASTContext().isSwiftVersionAtLeast(5)) {
+    SILLocation fnLoc = TheMemory.getFunction().getLocation();
+    if (auto *ctor = fnLoc.getAsASTNode<ConstructorDecl>()) {
```

Comment:
I'd rather we encode this in the MarkUninitializedInst than add more places in DI where we reach into the AST.